btx/processing/pipca.py

```diff
@@ -0,0 +1,1011 @@
+
+        if downsample:
+            if det_shape[-1] % bin_factor or det_shape[-2] % bin_factor:
+                print("Invalid bin factor, toggled off downsampling.")
+                self.downsample = False
+            else:
+                d = int(d / bin_factor**2)
+
+        return n, q, m, d
+
+    def run(self):
+        """
+        Perform iPCA on run subject to initialization parameters.
+        """
+        m = self.batch_size
+        num_images = self.num_images
+
+        # initialize and prime model, if specified
+        if self.priming and not self.benchmarking:
+            img_batch = self.get_formatted_images(
+                self.num_components, 0, self.num_features
+            )
+            self.prime_model(img_batch)
+        else:
+            self.U = np.zeros((self.split_counts[self.rank], self.num_components))
+            self.S = np.ones(self.num_components)
+            self.mu = np.zeros((self.split_counts[self.rank], 1))
+            self.total_variance = np.zeros((self.split_counts[self.rank], 1))
+
+        # divide remaining number of images into batches
+        # will become redundant in a streaming setting, need to change
+        rem_imgs = num_images - self.num_incorporated_images
+        batch_sizes = np.array(
+            [m] * np.floor(rem_imgs / m).astype(int)
+            + ([rem_imgs % m] if rem_imgs % m else [])
+        )
+
+        # update model with remaining batches
+        for batch_size in batch_sizes:
+            self.fetch_and_update_model(batch_size)
+
+        if self.benchmarking:
+            self.save_interval_data()
+
+    def get_formatted_images(self, n, start_index, end_index):
+        """
+        Fetch n - x image segments from run, where x is the number of 'dead' images.
+
+        Parameters
+        ----------
+        n : int
+            number of images to retrieve
+        start_index : int
+            start index of subsection of data to retrieve
+        end_index : int
+            end index of subsection of data to retrieve
+
+        Returns
+        -------
+        ndarray, shape (end_index-start_index, n-x)
+            n-x retrieved image segments of dimension end_index-start_index
+        """
+
+        bin_factor = self.bin_factor
+        downsample = self.downsample
+
+        # may have to rewrite eventually when number of images becomes large,
+        # i.e. streamed setting, either that or downsample aggressively
+        imgs = self.psi.get_images(n, assemble=False)
+
+        if downsample:
+            imgs = bin_data(imgs, bin_factor)
+
+        imgs = imgs[
+            [i for i in range(imgs.shape[0]) if not np.isnan(imgs[i : i + 1]).any()]
+        ]
+
+        num_valid_imgs, p, x, y = imgs.shape
+        formatted_imgs = np.reshape(imgs, (num_valid_imgs, p * x * y)).T
+
+        return formatted_imgs[start_index:end_index, :]
+
+    def prime_model(self, X):
+        """
+        Initialiize model on sample of data using batch PCA.
```

Comment:
Do we define what the model is anywhere? That'd be useful.

Reply:
No mention explicitly, I'm not sure of the best place to add clarify this since the model (self.U, self.S, self.mu, self.total_variance) aren't initialized in the constructor. Thoughts?